### PR TITLE
CB-6397: FreeIPA Single mode can't be repaired when FreeIPA host was …

### PR DESCRIPTION
**Pull request by holleyism**
Azure has a similar lifecycle to AWS and stopped instances cannot be
rebooted and must be started.

Manually verified against Azure instance.  With and without force in
running and stopped states.

### CODE REVIEW

**Comment by lacikaaa** [comment]
isn't this fix needed for rc-2.19?

**Comment by holleyism** [comment]
> isn't this fix needed for rc-2.19?

It's technically a bug for 2.19, but I thought it was a minor issue with a valid workaround and didn't want to impact 2.19.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java [hunk 1]
```diff
@@ -68,22 +68,47 @@ public String getConsoleOutput(AuthenticatedContext authenticatedContext, CloudI
...
+        String resourceGroupName = azureUtils.getResourceGroupName(ac.getCloudContext(), vm.getCloudInstance());
+        completables.add(azureClient.startVirtualMachineAsync(resourceGroupName, vm.getCloudInstance().getInstanceId())
+                .doOnError(throwable -> {
+                    LOGGER.error("Error happend on azure instance start: {}", vm, throwable);
+                    statuses.add(new CloudVmInstanceStatus(vm.getCloudInstance(), InstanceStatus.FAILED, throwable.getMessage()));
+                })
+                .doOnCompleted(() -> statuses.add(new CloudVmInstanceStatus(vm.getCloudInstance(), InstanceStatus.STARTED)))
+                .subscribeOn(Schedulers.io()));
+
+    }
```
Comment (lacikaaa) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java:110]: I think there is only one line difference in these 2 methods. caould you refactor it? I would do something like:
```
if (vm.getStatus() == InstanceStatus.STARTED) {
    doReboot(completables, vm, statuses, () -> azureClient.rebootVirtualMachineAsync(resourceGroupName, vm.getCloudInstance().getInstanceId()));
} else if (vm.getStatus() == InstanceStatus.STOPPED) {
    doReboot(completables, vm, statuses, () -> azureClient.startVirtualMachineAsync(resourceGroupName, vm.getCloudInstance().getInstanceId()));
}
```
and
```
private void doReboot(List<Completable> completables, CloudVmInstanceStatus vm, List<CloudVmInstanceStatus> statuses, Supplier<Completable> supplier) {
        completables.add(supplier.get()
                .doOnError(throwable -> {
                    LOGGER.error("Error happend on azure instance reboot: {}", vm, throwable);
                    statuses.add(new CloudVmInstanceStatus(vm.getCloudInstance(), InstanceStatus.FAILED, throwable.getMessage()));
                })
                .doOnCompleted(() -> statuses.add(new CloudVmInstanceStatus(vm.getCloudInstance(), InstanceStatus.STARTED)))
                .subscribeOn(Schedulers.io()));
    }
```

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java [hunk 2]
```diff
@@ -68,22 +69,35 @@ public String getConsoleOutput(AuthenticatedContext authenticatedContext, CloudI
...
+                LOGGER.error(String.format("Unable to reboot instance %s because of invalid status %s.",
+                        vm.getCloudInstance().getInstanceId(), vm.getStatus().toString()));
+            }
         }
-        Completable.merge(rebootCompletables).await();
+        Completable.merge(completables).await();
         return statuses;
     }
 
+    private void doReboot(List<Completable> completables, AuthenticatedContext ac, CloudVmInstanceStatus vm, List<CloudVmInstanceStatus> statuses,
```
Comment (lacikaaa) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java:90]: so `AuthenticatedContext ac` is not used anymore, could you remove?
also could you give `supplier` a bit more meaningful name? thanks

Comment (holleyism) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/AzureInstanceConnector.java:90]: fixed

**Pull request merged**